### PR TITLE
Documentation for dependency resolvers

### PR DESCRIPTION
**Pull request by pvanheus**
Documentation largely derived from my blog post (http://pvh.wp.sanbi.ac.za/2015/10/09/how-galaxy-resolves-dependencies-or-not/) on how dependency resolvers are and can be configured for Galaxy.

### CODE REVIEW

File: config/dependency_resolvers_conf.xml.sample [hunk 1]
```diff
@@ -1,7 +1,10 @@
 <dependency_resolvers>
 <!-- the default configuration, first look for dependencies installed from the toolshed -->
   <tool_shed_packages />
-<!-- then look for env.sh files profile according to the "galaxy packages" schema -->
+<!-- then look for env.sh files in directories according to the "galaxy packages" schema.
+     These resolvers can take a base_path attribute to specify where to look for
```
Comment (nsoranzo) [inline config/dependency_resolvers_conf.xml.sample:5]: I can't find the code where `base_path` attribute should be parsed, `lib/galaxy/tools/deps/__init__.py` seems to always use `config.tool_dependency_dir` .

File: doc/source/admin/dependency_resolvers.rst [hunk 2]
```diff
@@ -0,0 +1,143 @@
+Dependency Resolvers in Galaxy
+==============================
+
+There are two parts to building a link between Galaxy and command line bioinformatics tools: the tool XML that
+specifies a mapping between the Galaxy web user interface and the tool command line and tool dependencies that specify
+how to source the actual packages that implement the tool’s commands. The final script that Galaxy submits to run a
+the job uses includes commands, such as changes to the ``PATH`` environment variable, that are generated by *dependency
```
Comment (nsoranzo) [inline doc/source/admin/dependency_resolvers.rst]: Remove `the` at the start of the line (there's already an `a` at the end of the previous line) and `uses`.

File: doc/source/admin/dependency_resolvers.rst [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+    <requirement type="package" version="0.7.10.039ea20639">bwa</requirement>
+
+In some cases these requirement tags can be specified without a version
+
+.. code-block:: xml
+
+    <requirement type="package">bedtools</requirement>
+
+The requirement turn into inputs to the dependency resolver. Each dependency resolver is thus given given one or
```
Comment (nsoranzo) [inline doc/source/admin/dependency_resolvers.rst:24]: s/requirement/requirements/

File: doc/source/admin/dependency_resolvers.rst [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+two inputs: the name of the dependency to resolve and, in most cases, the version number of the
+dependency. By default, the dependency resolver configuration contains three items. First, the
```
Comment (nsoranzo) [inline doc/source/admin/dependency_resolvers.rst]: I'd move the part starting with "By default" to the `Default Dependency Resolvers` subsection.

File: doc/source/admin/dependency_resolvers.rst [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    <galaxy_packages versionless="true" />
+  </dependency_resolvers>
+
+Tool Shed Dependency Resolver
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``tool_shed_packages`` dependency resolver works with packages installed from the Galaxy Tool Shed. When a package
+is installed from the Tool Shed it creates a directory structure under the directory that is specified as the
+``tool_dependency_dir`` in Galaxy's configuration. This directory structure contains references to the tool's name,
+owner (in the Tool Shed) and version number (amongst other things) and ultimately contains a file name ``env.sh``
```
Comment (nsoranzo) [inline doc/source/admin/dependency_resolvers.rst]: s/name/named/

File: doc/source/admin/dependency_resolvers.rst [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+that contains commands to make the dependency runnable. This is installed, along with the packaged tool, by the tool
+package and doesn't require any configuration by the Galaxy administrator.
+
+The Tool Shed dependency resolver is not able to resolve package requirements that do not have a version number,
+like the `bedtools` example above.
+
+Galaxy Packages Dependency Resolver
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``galaxy_packages`` dependency resolver is allows Galaxy admins to specify how Galaxy should load manually
```
Comment (nsoranzo) [inline doc/source/admin/dependency_resolvers.rst]: Remove `is`.

File: doc/source/admin/dependency_resolvers.rst [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+installed packages. In the default configuration this is called twice, once immediately after
+the ``tool_shed_packages`` resolver and then a second time in its *versionless* mode.
```
Comment (nsoranzo) [inline doc/source/admin/dependency_resolvers.rst]: Actually, in the default configuration this is called 0, 1 or 2 times, depending on the previous resolvers succeeding or not. I'd just say that "This resolver can be configured either to use the version string or in _versionless_ mode.

File: doc/source/admin/dependency_resolvers.rst [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+    if [ -z "$MODULEPATH" ] ; then
+      . /etc/profile.d/module.sh
+    fi
+
+    module add bedtools/bedtools-2.20.1
+
+The Galaxy Package dependency resolver operates quite similarly when used in versionless module. Instead of looking
+for a directory named after a version, it looks for a directory ending in ``default``. For example
+``bedtools/default``. It then looks for a `bin` subdirectly or ``envh.sh`` and incorporates these in the tool
```
Comment (nsoranzo) [inline doc/source/admin/dependency_resolvers.rst]: s/subdirectly/subdirectory/

File: doc/source/admin/dependency_resolvers.rst [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    <requirement type="package" version="0.7.10.039ea20639">bwa</requirement>
+
+In some cases these requirement tags can be specified without a version
+
+.. code-block:: xml
+
+    <requirement type="package">bedtools</requirement>
+
+The requirement turn into inputs to the dependency resolver. Each dependency resolver is thus given given one or
+two inputs: the name of the dependency to resolve and, in most cases, the version number of the
```
Comment (nsoranzo) [inline doc/source/admin/dependency_resolvers.rst]: I would prefer `version string` to `version number` (here and in the following).

**Comment by nsoranzo** [comment]
Thanks @pvanheus, nice work!

File: config/dependency_resolvers_conf.xml.sample (hunk 1 above)
Comment (pvanheus) [inline config/dependency_resolvers_conf.xml.sample:5]: Its used in lib/galaxy/tools/deps/resolvers/resolver_mixins.py in the UsesToolDependencyDirMixin which is used by the galaxy_packages resolver.

File: doc/source/admin/dependency_resolvers.rst (hunk 3 above)
Comment (pvanheus) [inline doc/source/admin/dependency_resolvers.rst:24]: I'll make this change but the tag is, in fact, called requirement.

File: config/dependency_resolvers_conf.xml.sample (hunk 1 above)
Comment (nsoranzo) [inline config/dependency_resolvers_conf.xml.sample:5]: Thanks!

**Comment by jmchilton** [comment]
Utterly and wonderfully fantastic @pvanheus!

File: doc/source/admin/dependency_resolvers.rst (hunk 3 above)
Comment (nsoranzo) [inline doc/source/admin/dependency_resolvers.rst:24]: I think you misinterpreted my comment, you modified line 11 while I was referring to line 24: "The requirement turn into inputs..." should be "The requirements turn into inputs..."

**Pull request merged**